File: KeeperLegacyTests/CreatureCareTests.swift
import XCTest
@testable import KeeperLegacy

// MARK: - Creature Care Tests
// Tests for feeding, playing, cleaning, and stat decay.

final class CreatureCareTests: XCTestCase {

    func makeCreature(catalogID: String = "aquaburst") -> CreatureInstance {
        CreatureInstance(catalogID: catalogID)
    }

    // MARK: Feeding

    func testFeedIncreasesHunger() {
        var creature = makeCreature()
        creature.hunger = 0.2
        creature.feed(foodValue: 0.4)
        XCTAssertEqual(creature.hunger, 0.6, accuracy: 0.001)
    }

    func testFeedClampsAtMax() {
        var creature = makeCreature()
        creature.hunger = 0.9
        creature.feed(foodValue: 0.5)
        XCTAssertEqual(creature.hunger, 1.0)
    }

    func testFeedAlsoBoostsHappiness() {
        var creature = makeCreature()
        let beforeHappiness = creature.happiness
        creature.feed(foodValue: 0.5)
        XCTAssertGreaterThan(creature.happiness, beforeHappiness)
    }

    // MARK: Playing

    func testPlayIncreasesPlayfulness() {
        var creature = makeCreature()
        creature.playfulness = 0.3
        creature.play(toyName: "Ball", isFavoriteToy: false)
        XCTAssertGreaterThan(creature.playfulness, 0.3)
    }

    func testFavoriteToyGivesMoreHappiness() {
        var creature1 = makeCreature()
        var creature2 = makeCreature()
        creature1.happiness = 0.5
        creature2.happiness = 0.5

        creature1.play(toyName: "Random Toy", isFavoriteToy: false)
        creature2.play(toyName: "Favorite Toy", isFavoriteToy: true)

        XCTAssertGreaterThan(creature2.happiness, creature1.happiness)
    }

    func testFavoriteToyDiscovery() {
        var creature = makeCreature()
        XCTAssertFalse(creature.discoveredFavoriteToy)
        creature.play(toyName: "Bubble Wand", isFavoriteToy: true)
        XCTAssertTrue(creature.discoveredFavoriteToy)
    }

    func testFavoriteToyOnlyDiscoveredOnce() {
        var creature = makeCreature()
        let (_, discovered1) = creature.play(toyName: "Bubble Wand", isFavoriteToy: true)
        let (_, discovered2) = creature.play(toyName: "Bubble Wand", isFavoriteToy: true)
        XCTAssertTrue(discovered1)
        XCTAssertFalse(discovered2)   // Second play — already discovered
    }

    // MARK: Cleaning

    func testCleanSetsCleanlinessToFull() {
        var creature = makeCreature()
        creature.cleanliness = 0.1
        creature.clean()
        XCTAssertEqual(creature.cleanliness, 1.0)
    }

    // MARK: Overall Happiness

    func testOverallHappinessIsAverageOfStats() {
        var creature = makeCreature()
        creature.hunger      = 1.0
        creature.happiness   = 1.0
        creature.cleanliness = 1.0
        creature.affection   = 1.0
        creature.playfulness = 1.0
        XCTAssertEqual(creature.overallHappiness, 1.0, accuracy: 0.001)
    }

    func testSellMultiplierAtNeutral() {
        var creature = makeCreature()
        creature.hunger      = 0.5
        creature.happiness   = 0.5
        creature.cleanliness = 0.5
        creature.affection   = 0.5
        creature.playfulness = 0.5
        XCTAssertEqual(creature.sellMultiplier, 1.5, accuracy: 0.001)  // 1.0 + 0.5
    }

    // MARK: Stat Decay

    func testDecayReducesStats() {
        var creature = makeCreature()
        let beforeHunger = creature.hunger
        creature.applyDecay(hoursPassed: 10)
        XCTAssertLessThan(creature.hunger, beforeHunger)
    }

    func testDecayNeverGoesBelowZero() {
        var creature = makeCreature()
        creature.applyDecay(hoursPassed: 10000)
        XCTAssertGreaterThanOrEqual(creature.hunger,      0.0)
        XCTAssertGreaterThanOrEqual(creature.happiness,   0.0)
        XCTAssertGreaterThanOrEqual(creature.cleanliness, 0.0)
        XCTAssertGreaterThanOrEqual(creature.affection,   0.0)
        XCTAssertGreaterThanOrEqual(creature.playfulness, 0.0)
    }

    func testHungerDecaysFasterThanAffection() {
        var creature = makeCreature()
        creature.hunger    = 1.0
        creature.affection = 1.0
        creature.applyDecay(hoursPassed: 5)
        // Hunger decay rate is 1.5x, affection is 0.5x
        let hungerLost    = 1.0 - creature.hunger
        let affectionLost = 1.0 - creature.affection
        XCTAssertGreaterThan(hungerLost, affectionLost)
    }
}

File: KeeperLegacyGodot/Tests/CreatureCareTests.cs
// Tests/CreatureCareTests.cs
// Port of CreatureCareTests.swift
// Tests feeding, playing, cleaning, and stat decay on CreatureInstance.

using NUnit.Framework;
using KeeperLegacy.Models;

namespace KeeperLegacy.Tests
{
    [TestFixture]
    public class CreatureCareTests
    {
        private CreatureInstance MakeCreature(string catalogId = "aquaburst")
            => new CreatureInstance(catalogId);

        // ── Feeding ───────────────────────────────────────────────────────────

        [Test]
        public void FeedIncreasesHunger()
        {
            var c = MakeCreature();
            c.Hunger = 0.2;
            c.Feed(0.4);
            Assert.That(c.Hunger, Is.EqualTo(0.6).Within(0.001));
        }

        [Test]
        public void FeedClampsAtMax()
        {
            var c = MakeCreature();
            c.Hunger = 0.9;
            c.Feed(0.5);
            Assert.That(c.Hunger, Is.EqualTo(1.0));
        }

        [Test]
        public void FeedAlsoBoostsHappiness()
        {
            var c = MakeCreature();
            double before = c.Happiness;
            c.Feed(0.5);
            Assert.That(c.Happiness, Is.GreaterThan(before));
        }

        // ── Playing ───────────────────────────────────────────────────────────

        [Test]
        public void PlayIncreasesPlayfulness()
        {
            var c = MakeCreature();
            c.Playfulness = 0.3;
            c.Play("Ball", isFavoriteToy: false);
            Assert.That(c.Playfulness, Is.GreaterThan(0.3));
        }

        [Test]
        public void FavoriteToyGivesMoreHappiness()
        {
            var c1 = MakeCreature();
            var c2 = MakeCreature();
            c1.Happiness = 0.5;
            c2.Happiness = 0.5;

            c1.Play("Random Toy",   isFavoriteToy: false);
            c2.Play("Favorite Toy", isFavoriteToy: true);

            Assert.That(c2.Happiness, Is.GreaterThan(c1.Happiness));
        }

        [Test]
        public void FavoriteToyDiscovery()
        {
            var c = MakeCreature();
            Assert.That(c.DiscoveredFavoriteToy, Is.False);
            c.Play("Bubble Wand", isFavoriteToy: true);
            Assert.That(c.DiscoveredFavoriteToy, Is.True);
        }

        [Test]
        public void FavoriteToyOnlyDiscoveredOnce()
        {
            var c = MakeCreature();
            var (_, discovered1) = c.Play("Bubble Wand", isFavoriteToy: true);
            var (_, discovered2) = c.Play("Bubble Wand", isFavoriteToy: true);
            Assert.That(discovered1, Is.True);
            Assert.That(discovered2, Is.False);
        }

        // ── Cleaning ──────────────────────────────────────────────────────────

        [Test]
        public void CleanSetsCleanlinessToFull()
        {
            var c = MakeCreature();
            c.Cleanliness = 0.1;
            c.Clean();
            Assert.That(c.Cleanliness, Is.EqualTo(1.0));
        }

        // ── Overall Happiness ─────────────────────────────────────────────────

        [Test]
        public void OverallHappinessIsAverageOfStats()
        {
            var c = MakeCreature();
            c.Hunger = c.Happiness = c.Cleanliness = c.Affection = c.Playfulness = 1.0;
            Assert.That(c.OverallHappiness, Is.EqualTo(1.0).Within(0.001));
        }

        [Test]
        public void SellMultiplierAtNeutral()
        {
            var c = MakeCreature();
            c.Hunger = c.Happiness = c.Cleanliness = c.Affection = c.Playfulness = 0.5;
            Assert.That(c.SellMultiplier, Is.EqualTo(1.5).Within(0.001));
        }

        // ── Stat Decay ────────────────────────────────────────────────────────

        [Test]
        public void DecayReducesStats()
        {
            var c = MakeCreature();
            double before = c.Hunger;
            c.ApplyDecay(hoursPassed: 10);
            Assert.That(c.Hunger, Is.LessThan(before));
        }

        [Test]
        public void DecayNeverGoesBelowZero()
        {
            var c = MakeCreature();
            c.ApplyDecay(hoursPassed: 10000);
            Assert.That(c.Hunger,      Is.GreaterThanOrEqualTo(0.0));
            Assert.That(c.Happiness,   Is.GreaterThanOrEqualTo(0.0));
            Assert.That(c.Cleanliness, Is.GreaterThanOrEqualTo(0.0));
            Assert.That(c.Affection,   Is.GreaterThanOrEqualTo(0.0));
            Assert.That(c.Playfulness, Is.GreaterThanOrEqualTo(0.0));
        }

        [Test]
        public void HungerDecaysFasterThanAffection()
        {
            var c = MakeCreature();
            c.Hunger    = 1.0;
            c.Affection = 1.0;
            c.ApplyDecay(hoursPassed: 5);
            double hungerLost    = 1.0 - c.Hunger;
            double affectionLost = 1.0 - c.Affection;
            Assert.That(hungerLost, Is.GreaterThan(affectionLost));
        }
    }
}
